### ISSUE TITLE
weak volume locking during partitioning.

### ISSUE THREAD

**Issue by sgotti** [coreos/torus#230]
Actually volumes are locked using an etcd lease. If, for any reason, a client (like `torusblk`) is partitioned (at least it cannot talk to etcd) its lease will expire in 30 seconds, after this another client can start using the volume. If the first client become unpartitioned it'll restart working on the volume. At the end we'll have multiple client writing to the same volume causing data corruption (for example if using a local fs like ext4 on the volume like done with the k8s flex volume plugin).

I have a lot to learn on torus to suggest the best way to avoid this (so forgive me if I'm using some wrong terminology) but these are my 2 cents:

* Register a persistent lock (not a lease) on a volume in the mds.
This will require the client removing the lock, if this isn't done there's the need of a manual step to remove the lock or no one will be able to use it. 
This manual step is cumbersome since the user needs to ensure that the client (host or just the client process) is really down or terminate it before removing the lock.
This is something similar to what's done by ceph rbd lock/unlock commands and these are used by the k8s volume plugin to handle the partitioning cases (and they need manual intervention when a lock isn't removed)

* Use some forms of fencing.
Two examples:
 1. On the client (like `torusblk`) If a client loses its lease any subsequent action on a volume should be blocked. I'm not sure if this is feasible since every client operation should go to the mds to check if it has access and this will probably add too much overhead and slowdowns. Checking this only at a specified interval (or with a watch) will leave a (small) window where two concurrent operation can happen.
 2. On the storage nodes (`torusd`).
Every client will be unique identified in some ways and for every operation the storage node will check that the client can access the volume before accepting any operation. As the above I'm not sure if this is feasible since every operation will need to check this probably going to the mds for every operation.

**Comment by barakmich** [comment]
No operations on the metadata for a block device are allowed if the lock isn't present, so it will never have multiple clients writing to the same volume at once. However, operations done on the client that was partitioned, during the partition, may be lost.

So, there's a couple of decent options:

1) Your first option, a persistent lock that is potentially cumbersome but correct. 
2) Better reconnections. As long as the lease is alive, the lock is "persistent" -- so if we reconnect in 30 seconds, everything is fine.  This is okay, but having an appropriate way of failing client operations when etcd cannot be reached for sufficient amounts of time (and therefore the lock may be gone) is probably the right approach. Furthermore, all clients, seeing that their client lease no longer exists, should close all pending operations and grab a new lease. Cleaning up this codepath is a major step for stability.

**Comment by sgotti** [comment]
> No operations on the metadata for a block device are allowed if the lock isn't present, so it will never have multiple clients writing to the same volume at once.

@barakmich Great! Thanks a lot for the explanation. Now that I had time to look more deeply in the code I found it (`block/etcd.go` the `blocklock` key is created with the lease and used in the `SyncINode` and `DeleteVolume` functions).

Should I close this issue and open a new one for this enhancement or can we keep this (perhaps with a better title?).

> 2) Better reconnections.

Given the above I think this is the best solution.

I also noticed that, while the old client cannot sync the metadata, it'll continue writing blocks to the stores. Usually they are GCed but, only when the client with the expired lease is active (no other clients with good lease working on the volume), these block are never garbage collected. This is cleaned when a new client starts writing to the volume. Probably this will be fixed implementing this option. Should I open a new issue anyway?
